### PR TITLE
Added contributing guide, early 9.4 support

### PR DESCRIPTION
**Pull request by coot**
- Added LICENSE and NOTICE files
- Added CONTRIBUTING.md document
- Early support for GHC 9.4.2
- Use GHC 9.2.4 in CI

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# Contributing to io-sim
+
+We designed `io-classes` to be as close as possible to what `base` package
+provides.  Almost all `IO` instances intstantiate with api provided by one of
```
Comment (bolt12) [inline CONTRIBUTING.md:4]: I am not entirely sure what you mean with the second sentence.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+the core packages.  Please keep this in mind when adding new functionality.
+
+## Style Guide
+
+Please follow the [style guide][style-guide].
+
+## Pull Request
+
+Ideally the series of commits you want to include should be written with the
+revier in mind.  This will make it easier for the maintainers to review your
```
Comment (bolt12) [inline CONTRIBUTING.md:14]: typo: revier -> reviewer

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+changes.  Each commit should heave a clear explanation what it changes and most
+importantly why the changes are necessary.   We prefer if commits are small and
+just introduce a single change.  For example if you are adding a new typeclass
+in `io-classes` and you want `io-sim` and `io-sim-por` to support it, we'd
+expect at least three commits: first which introduces the new type class, and
+seprate commits which add implement it for `IOSim` and `IOSimPOR`, since both
```
Comment (bolt12) [inline CONTRIBUTING.md:20]: seprate -> separate

File: io-sim/src/Control/Monad/IOSim/Internal.hs [hunk 4]
```diff
@@ -1046,17 +1046,17 @@ saveTVar TVar{tvarCurrent, tvarUndo} = do
 revertTVar :: TVar s a -> ST s ()
 revertTVar TVar{tvarCurrent, tvarUndo} = do
     -- pop the undo stack, and revert the current value
-    (v:vs) <- readSTRef tvarUndo
-    !_ <- writeSTRef tvarCurrent v
-    !_ <- writeSTRef tvarUndo    vs
+    vs <- readSTRef tvarUndo
+    !_ <- writeSTRef tvarCurrent (head vs)
+    !_ <- writeSTRef tvarUndo    (tail vs)
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:1051]: Why this change? I actually prefer how it was previously. you can rename `(v:vs)` to `(head:tail)` and recover readability

File: io-sim/src/Control/Monad/IOSim/Internal.hs [hunk 5]
```diff
@@ -1046,17 +1046,17 @@ saveTVar TVar{tvarCurrent, tvarUndo} = do
...
     return ()
 {-# INLINE revertTVar #-}
 
 commitTVar :: TVar s a -> ST s ()
 commitTVar TVar{tvarUndo} = do
+    vs <- readSTRef tvarUndo
     -- pop the undo stack, leaving the current value unchanged
-    (_:vs) <- readSTRef tvarUndo
-    !_ <- writeSTRef tvarUndo vs
+    !_ <- writeSTRef tvarUndo (tail vs)
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSim/Internal.hs:1059]: Here as well

File: io-sim/src/Control/Monad/IOSimPOR/Internal.hs [hunk 6]
```diff
@@ -1264,16 +1264,16 @@ saveTVar TVar{tvarCurrent, tvarUndo} = do
 revertTVar :: TVar s a -> ST s ()
 revertTVar TVar{tvarCurrent, tvarUndo} = do
     -- pop the undo stack, and revert the current value
-    (v:vs) <- readSTRef tvarUndo
-    writeSTRef tvarCurrent v
-    writeSTRef tvarUndo    vs
+    vs <- readSTRef tvarUndo
+    writeSTRef tvarCurrent (head vs)
+    writeSTRef tvarUndo    (tail vs)
```
Comment (bolt12) [inline io-sim/src/Control/Monad/IOSimPOR/Internal.hs:1269]: Here too

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 4 above)
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:1051]: Also maybe these changes could go into a different commit?

File: io-classes/src/Control/Monad/Class/MonadMVar.hs [hunk 7]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE DefaultSignatures      #-}
 {-# LANGUAGE QuantifiedConstraints  #-}
 {-# LANGUAGE TypeFamilyDependencies #-}
+{-# LANGUAGE TypeOperators          #-}
```
Comment (bolt12) [inline io-classes/src/Control/Monad/Class/MonadMVar.hs:4]: Why this addition?

**Review by bolt12 — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 1 above)
Comment (coot) [reply CONTRIBUTING.md:4]: I wanted to convey they sprit of such insetances as [this one](https://github.com/input-output-hk/io-sim/blob/main/io-classes/src/Control/Monad/Class/MonadSTM.hs?plain=1#L410-L446).

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 4 above)
Comment (coot) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:1051]: `ST` has no `MonadFail` instance (`ghc-9.4.2`)

File: io-classes/src/Control/Monad/Class/MonadMVar.hs (hunk 7 above)
Comment (coot) [reply io-classes/src/Control/Monad/Class/MonadMVar.hs:4]: `ghc-9.4` warns about using `~` without `TypeOperators` enabled.

File: io-sim/src/Control/Monad/IOSim/Internal.hs (hunk 4 above)
Comment (bolt12) [reply io-sim/src/Control/Monad/IOSim/Internal.hs:1051]: Oh I see!!

**Review by bolt12 — APPROVED** [review]

**Comment by coot** [comment]
These are the settings one needs to add to `cabal.project.local` file right now to build it with `ghc-9.4`.
```
allow-newer: typed-protocols-cborg:base
           , typed-protocols-examples:base
           , cborg:base
           , cborg:ghc-prim
           , cborg:primitive
           , serialise:ghc-prim
           , serialise:time
           , serialise:text
           , serialise:bytestring
```
